scipy/linalg/tests/test_lapack.py

```diff
@@ -1645,3 +1645,50 @@ def test_getc2_gesc2():
         else:
             assert_array_almost_equal(desired_cplx.astype(dtype),
                                       x/scale, decimal=4)
+
+
+def test_gttrf_gttrs():
+    np.random.seed(42)
+    n = 10
+    for index, dtype in enumerate(DTYPES):
+        atol = 10**-(np.finfo(dtype).precision-1)
+        if index < 2:
+            du = np.random.rand(n-1)
+            d = np.random.rand(n)
+            dl = np.random.rand(n-1)
+        else:
+            du = np.random.rand(n-1) * 1j
+            d = np.random.rand(n) * 1j
+            dl = np.random.rand(n-1) * 1j
+
+        diag_cpy = np.array([dl.copy(), d.copy(), du.copy()])
+
+        A = np.diag(d) + np.diag(dl, -1) + np.diag(du, 1)
+        x = np.random.rand(n)
+        b = A @ x
+
+        dgttrf = get_lapack_funcs('gttrf', dtype=dtype)
+        dgttrs = get_lapack_funcs('gttrs', dtype=dtype)
+
+        _dl, _d, _du, du2, ipiv, info = dgttrf(dl, d, du)
+
```

Comment:
Let's add the test that the factors multiply to reproduce the (permuted) original matrix.

Reply:
Haven't finished this yet. It should be as simple as what we did to permuting A to get the new factors, yes?

Reply:
@mdhaber What factor is this missing from this? I think that the permutation order is being done correctly on the original matrix, but A permuted is not matching the combined dl, d, du outputs of gttrf. Is there additional swapping / eliminations that are being done outside of IPIV with du2?

Reply:
The input `dl`, `d`, `du` describe the original matrix `A` as you have written, `np.diag(dl, -1) + np.diag(d, 0) + np.diag(du, 1)`.
The output variables `_dl`, `_d`, `_du`, and `_du2` also describe `A` but in a different way. 
`_dl` is the -1 diagonal of a lower triangular matrix `L`, so `L = np.eye(n) + np.diag(dl, -1)`.
`_d`, `_du`, and `_du2` are the first three diagonals of an upper triangular matrix `U`: `U = np.diag(_d) + np.diag(_du, 1) + np.diag(_du2, 2)`.
Permuted `A` is the product of `L` and `U`, so `assert_allclose(L@U, A[perm_fnl], rtol=rtol)`.
I think.

Reply:
Hmm, I am seeing negative numbers inside the result of L@U despite A originally having none. I let L be `np.eye(n) + np.diag(_dl, -1)` and U be `np.diag(_d, 0) + np.diag(_du, 1) + np.diag(du2, 2)`, but `L@U` gives a couple values that are negative, so even before permuting it seems that something still isn't right here.

I saw some talk of multipliers on the  [NAG Manual example result](https://www.nag.com/numeric/fl/nagdoc_latest/examples/baseresults/f07cdfe.r.html) page, are those relevant here?

Not much new code other than what's above, so I didn't add any commits for this.

Reply:
I see. Well, I'm pretty certain that `U` is right because the documentation is very explicit, e.g. 

> On exit, D is overwritten by the n diagonal elements of the upper triangular matrix U from the LU factorization of A.

and that's what `U = np.diag(_d)...` does.

But the documentation about `dl` does not say "subdiagonal" of `L`:

> On exit, DL is overwritten by the (n-1) multipliers that define the matrix L from the LU factorization of A.

I assumed it meant subdiagonal, but it says "multipliers". I'm trying to find what that means.

Reply:
I see. I read that article, and even though I'm not 100% on it, I think I see the general idea of what is going on with what you provided. It looks like A - L@U brings all of the entries in the matrix to near zero. 

However, I've been running into an issue with how assert_equals compares A and L@U. For example, if the two elements from the matrix being compared were` 0.` and `6.059189e-09, ` with the way troll works, I think that it won't end up passing as true. I saw this: `absolute(a - b) <= (atol + rtol * absolute(b))` in the documentation, so I thought that possibly the problem would be `b` was ending up as zero and therefore the tolerance was zero. But, I think since either `a` or `b` has to be zero, modifying rtol here might not work. Is it possible that I can change atol instead since it it value independent?

I'm working on adapting it to work within the for loop for complex numbers as well.

Reply:
Yes, it's ok to change `atol` too, if you need to.

Reply:
It seems that `atol` was 0, which is why this wasn't passing. Go ahead and set a reasonable `atol`.